### PR TITLE
applyPatches.sh options: --clean --minimal --minclang

### PR DESCRIPTION
**Pull request by foresto**
The lineageos_lilac README mentions that applying all patches is "only for advanced users", but doesn't offer another way to get a working build short of manually identifying and applying a handful of required patches.

I wanted to solve this, so a non-advanced build would be simpler and less error-prone, so I added command line options to applyPatches.sh:

- `--minimal` : Apply only the patches that are required for a build to succeed, plus security patches. Right now the only required one is fix-custom-apn-script.patch, without which the build fails with `ValueError: list.remove(x): x not in list`. I imagine there could be more required patches in the future.
- `--minclang` : Apply the minimal & security patches plus those required for the new clang version, as described in the lineageos_lilac README.
- `--clean` : Revert patched project directories to their unpatched state. I found this helpful when your recent ASB patches arrived, because they seem to merge several months of the upstream patches together, and therefore conflicted with my dirs that already had the March patches applied. This option makes it easy to roll already-applied ones back in preparation for a new month's merged patches, and is much faster than waiting hours for a fresh repo init/sync.

I've only done this for the lineage-17.1 branch, since that's the only LineageOS version I'm using.

One possible drawback I see here is that applyPatches.sh embeds the list of required patches, so it necessarily diverges between the various lineage-xx.x branches. Perhaps that's fine. 

A different approach might be to move the patches into the lineageos_lilac project, and put each version's patches into subdirs, like that project already does with manifests. Then maybe add a list of each LineageOS version's required patches in a separate file, and let applyPatches.sh look them up instead of embedding them in the script. That would allow a generic applyPatches.sh script to work for all versions.

This PR is mainly to share what I'm doing in my own fork, in case it's either useful enough to be accepted as-is, or as inspiration for something else.

### CODE REVIEW

**Comment by foresto** [comment]
Here's the matching change to the lineageos_lilac README: 

https://github.com/foresto/lineageos_lilac/commit/061cd3caea9a5ea9ee72732632699eb55255ffa0

File: patches/applyPatches.sh [hunk 1]
```diff
@@ -23,6 +23,27 @@ if [ ! -d "$repo_root/device/sony/lilac/patches" ]; then
   showError "Failed to find repository root at $repo_root"
 fi
 
+patch_set=all
+script_name=$(basename "$0")
+for arg in "$@"; do
+    shift
```
Comment (Flamefire) [inline patches/applyPatches.sh]: Wrong order: Only a single shift is required
```suggestion
shift
for arg in "$@"; do
```

File: patches/applyPatches.sh [hunk 2]
```diff
@@ -35,6 +56,18 @@ function applyPatch {
         showError "Faulty patch: $patch"
     fi
 
+    if [ "${clean_patch_dirs:-}" ]; then
+        if ! git -C "$patch_dir" diff --quiet; then
+            echo -en "Cleaning ${patch_dir}: "
```
Comment (Flamefire) [inline patches/applyPatches.sh]:
```suggestion
            echo -n "Cleaning ${patch_dir}: "
```

**Review by Flamefire — CHANGES_REQUESTED** [review]
Thanks, I can accept that as the default is kept unchanged. Only 2 nits to fix please, you can simply accept them in the Github UI.

**Comment by foresto** [comment]
Actually, both those nits were remnants from an earlier revision and are no longer needed. Thanks for catching them. Fixed.

**Comment by Flamefire** [comment]
True and the `shift` would have been even wrong, so all good now.

BTW: For boolean options I see an advantage of setting the variable to `0` before the loop and testing with e.g. `if ((clean_patch_dirs == 1)); then`, currently it might pick up an exported variable from outside and the check `[ "${clean_patch_dirs:-}" ]` requires knowledge about variable substitution and how strings are evaluated in conditions, e.g. that the value  doesn't matter, which can make it easily fail later when a default is introduced, e.g. `clean_patch_dirs="0"` or so would yield true.

**Comment by foresto** [comment]
Fair enough. Something like #21?

**Comment by Flamefire** [comment]
Yes. Although you didn't have to do that change, I just wanted to show an alternative in this context

**Pull request merged**